src/core/net/ip6.cpp

```diff
@@ -998,13 +998,26 @@ Error Ip6::ProcessReceiveCallback(Message           &aMessage,
                                   Message::Ownership aMessageOwnership)
 {
     Error    error   = kErrorNone;
-    Message *message = &aMessage;
+    Message *message = nullptr;
 #if OPENTHREAD_CONFIG_IP6_BR_COUNTERS_ENABLE
     Header header;
 
     IgnoreError(header.ParseFrom(aMessage));
 #endif
 
+    // `message` points to the `Message` instance we own in this
+    // method. If we can take ownership of `aMessage`, we use it as
+    // `message`. Otherwise, we may create a clone of it and use as
+    // `message`. `message` variable will be set to `nullptr` if the
+    // message ownership is transferred to an invoked callback. At
+    // the end of this method we free `message` if it is not `nullptr`
+    // indicating it was not passed to a callback.
+
+    if (aMessageOwnership == Message::kTakeCustody)
+    {
+        message = &aMessage;
```

Comment:
Can we move this assignment to the switch statement below?

Reply:
Seems we can't.
The VerifyOrExit below may fail and aMessage won't be released.

However, seems we can delete the `if` directly since we will free `*message` anyway?

Reply:
As @erjiaqing said, we cannot.

- We should set it at the top of function and before any check that may exit early.
- When `aMessageOwnership` is `kTakeCustody`, then we own the passed-in `aMessage` and should either pass its ownership to a callback or free it ourself. 
- There are a bunch of `VerifyOrExit()` checks which can exit early before we get to `switch()` statement.